### PR TITLE
add afterimage effect

### PR DESCRIPTION
**Pull request by HypnosNova**
A postprocessing of afterimage effect. 
Looks pretty cool!

### CODE REVIEW

File: examples/js/postprocessing/AfterimagePass.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+	this.textureComp = new THREE.WebGLRenderTarget( window.innerWidth, window.innerHeight, {
+
+		minFilter: THREE.LinearFilter,
+		magFilter: THREE.NearestFilter,
+		format: THREE.RGBAFormat
+
+	} );
+
+    this.textureOld = new THREE.WebGLRenderTarget( window.innerWidth, window.innerHeight, {
```
Comment (Mugen87) [inline examples/js/postprocessing/AfterimagePass.js]: Please ensure that the formatting of your files are correct.

File: examples/js/postprocessing/AfterimagePass.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+	} );
+
+	this.shaderMaterial = new THREE.ShaderMaterial( {
+
+		uniforms: this.uniforms,
+		vertexShader: this.shader.vertexShader,
+		fragmentShader: this.shader.fragmentShader
+
+    } );
```
Comment (Mugen87) [inline examples/js/postprocessing/AfterimagePass.js]: Format

File: examples/js/shaders/AfterimageShader.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+			"vec4 texelOld = texture2D( tOld, vUv );",
+			"vec4 texelNew = texture2D( tNew, vUv );",
+			
+			"texelOld *= damp;",
+			
+			"texelOld.r *= when_gt( texelOld.r, 0.2 );",
+    		"texelOld.g *= when_gt( texelOld.g, 0.2 );",
+    		"texelOld.b *= when_gt( texelOld.b, 0.2 );",
+    		"texelOld.a *= when_gt( texelOld.a, 0.2 );",
```
Comment (Mugen87) [inline examples/js/shaders/AfterimageShader.js]: Format

File: examples/webgl_postprocessing_afterimage.html [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+			function animate() {
+
+				requestAnimationFrame( animate );
+
+				var time = Date.now();
+
+				mesh.rotation.x += 0.005;
+				mesh.rotation.y += 0.01;
+				
+				if(openPass){
```
Comment (Mugen87) [inline examples/webgl_postprocessing_afterimage.html]: Code style issue. It should be `if ( openPass ) {`

File: examples/webgl_postprocessing_afterimage.html [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+					composer.render();
+
+				} else {
+
+					renderer.render(scene, camera);
```
Comment (Mugen87) [inline examples/webgl_postprocessing_afterimage.html]: `renderer.render( scene, camera );`

File: examples/webgl_postprocessing_afterimage.html [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+		<script src="js/postprocessing/AfterimagePass.js"></script>
+		
+		<script src="js/libs/dat.gui.min.js" type="text/javascript" charset="utf-8"></script>
+
+		<script>
+
+			var camera, scene, renderer, composer;
+			var mesh, light;
+
+			var afterimagePass, openPass=true;
```
Comment (Mugen87) [inline examples/webgl_postprocessing_afterimage.html]: What does `openPass` mean?

**Review by Mugen87 — CHANGES_REQUESTED** [review]

File: examples/webgl_postprocessing_afterimage.html (hunk 6 above)
Comment (HypnosNova) [reply examples/webgl_postprocessing_afterimage.html]: i renamed it to turnOnPass, just mean to switch normal rendering result and postprocessing result

File: examples/webgl_postprocessing_afterimage.html [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+		<script src="js/postprocessing/AfterimagePass.js"></script>
+		
+		<script src="js/libs/dat.gui.min.js" type="text/javascript" charset="utf-8"></script>
+
+		<script>
+
+			var camera, scene, renderer, composer;
+			var mesh, light;
+
+			var afterimagePass, turnOnPass = true;
```
Comment (Mugen87) [inline examples/webgl_postprocessing_afterimage.html]: I would call it just `enable` so it's consistent to: https://threejs.org/examples/#webgl_postprocessing_sobel

Please also update the label of the `dat.gui` interface.

File: examples/webgl_postprocessing_afterimage.html (hunk 6 above)
Comment (Mugen87) [reply examples/webgl_postprocessing_afterimage.html]: I would call it just `enable` so it's consistent to: https://threejs.org/examples/#webgl_postprocessing_sobel

Please also update the label of the `dat.gui` interface.

**Comment by Mugen87** [comment]
Great. Can you also add your example to the following file:

https://github.com/mrdoob/three.js/blob/dev/examples/files.js#L223

Otherwise it won't be visible in the menu.

File: examples/js/shaders/AfterimageShader.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
+/**
+ * @author HypnosNova / https://www.threejs.org.cn/gallery/
+ * I'm lucky to get this domain name.(threejs.org.cn)
```
Comment (Mugen87) [inline examples/js/shaders/AfterimageShader.js]: Can you please remove this sentence? :wink:

File: examples/js/shaders/AfterimageShader.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+ *
+ * Afterimage shader
+ * I created this effect inspired by a demo on codepen. (Unfortunately, I forgot the url)
```
Comment (Mugen87) [inline examples/js/shaders/AfterimageShader.js]: It would be actually good if the example refers to the original code/author.

**Review by Mugen87 — APPROVED** [review]

File: examples/js/shaders/AfterimageShader.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+			"vec4 texelOld = texture2D( tOld, vUv );",
+			"vec4 texelNew = texture2D( tNew, vUv );",
+			
+			"texelOld *= damp;",
+			
+			"texelOld.r *= when_gt( texelOld.r, 0.2 );",
+			"texelOld.g *= when_gt( texelOld.g, 0.2 );",
+			"texelOld.b *= when_gt( texelOld.b, 0.2 );",
+			"texelOld.a *= when_gt( texelOld.a, 0.2 );",
```
Comment (WestLangley) [inline examples/js/shaders/AfterimageShader.js]: Try this, instead
```glsl
"vec4 when_gt( vec4 x, float y ) {",
...
"texelOld *= when_gt( texelOld, 0.2 );",
```

**Comment by WestLangley** [comment]
Wireframe `MeshNormalMaterial` is a bit unusual.

Is this designed to work only with wireframes? This is `MeshNormalMaterial` with wireframe false:

screen shot 2018-04-22 at 2 08 16 pm

File: examples/js/shaders/AfterimageShader.js (hunk 10 above)
Comment (HypnosNova) [reply examples/js/shaders/AfterimageShader.js]: thanks, i've approved the code and fixed the effect problem. Now it looks good without wireframe

Comment (HypnosNova) [reply examples/js/shaders/AfterimageShader.js]: @WestLangley thanks, i've approved the code and fixed the effect problem. Now it looks good without wireframe.

**Comment by Mugen87** [comment]
Yes. If we add such a shader to the repo, it should have a generic implementation and a stable parameterization.

**Comment by mrdoob** [comment]
Live: https://rawgit.com/HypnosNova/three.js/afterimage-effect/examples/webgl_postprocessing_afterimage.html

**Comment by mrdoob** [comment]
Wouldn't this be a good replacement of `webgl_trails.html`?
More stable than relying on WebGL's `preserveDrawingBuffer` which has bugs in some platforms and seems to be a headache to maintain.

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**